cpp/ql/src/semmle/code/cpp/controlflow/StackVariableReachability.qll

```diff
@@ -80,46 +324,185 @@ abstract class StackVariableReachability extends string {
         j > i and
         sink = bb.getNode(j) and
         isSink(sink, v) and
-        not exists(int k | isBarrier(bb.getNode(k), v) | k in [i + 1 .. j - 1])
+        not isBarrier(bb.getNode(pragma[only_bind_into]([i + 1 .. j - 1])), v)
       )
       or
       not exists(int k | isBarrier(bb.getNode(k), v) | k > i) and
-      bbSuccessorEntryReaches(bb, v, sink, _)
+      bbSuccessorEntryReaches(source, bb, v, sink, _)
+    )
+  }
+
+  private Condition getASinkCondition(SemanticStackVariable v) {
+    exists(BasicBlock bb |
+      revBbEntryReachesLocally(bb, v, _, this) and
+      result.getCondition().controls(bb, result.getTruthValue())
+    )
+  }
+
+  private Condition getABarrierCondition(SemanticStackVariable v) {
+    exists(BasicBlock bb |
+      isBarrier(bb.getANode(), v) and
+      result.getCondition().controls(bb, result.getTruthValue())
+    )
+  }
+
+  /**
+   * Gets a condition with a known truth value in `bb` when the control-flow starts at the source
+   * node `source` and we're tracking reachability using variable `v` (that is,
+   * `this.isSource(source, v)` holds).
+   *
+   * This predicate is `pragma[noopt]` as it seems difficult to get the correct join order for the
+   * recursive case otherwise:
+   * revBbSuccessorEntryReaches0(bb) -> getASuccessor -> prev_delta ->
+   * revBbSuccessorEntryReaches0(pred) -> {isLoopCondition, isDisjunctionCondition, isLoopVariantCondition}
+   */
+  pragma[noopt]
```

Comment:
That is true. I didn't talk to them explicitly about this. I'll raise it with the core team next time I have a chat with them. It should be possible to avoid the `noopt`, but I couldn't find the right combination of `only_bind_into` and `only_bind_out` that gave the order I wrote in the comments. I don't think it should block this PR, though.

Reply:
I agree